File: psicogest/backend/app/schemas/rips.py
"""Pydantic schemas for RIPS export endpoints."""
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class RipsGenerateRequest(BaseModel):
    year: int = Field(..., ge=2020, le=2030)
    month: int = Field(..., ge=1, le=12)


class RipsValidateRequest(BaseModel):
    year: int = Field(..., ge=2020, le=2030)
    month: int = Field(..., ge=1, le=12)


class RipsErrorItem(BaseModel):
    session_id: str | None = None
    field: str
    value: str | None = None
    message: str


class RipsWarningItem(BaseModel):
    session_id: str | None = None
    field: str
    value: str | None = None
    message: str


class RipsValidateResponse(BaseModel):
    valid: bool
    errors: list[RipsErrorItem]
    warnings: list[RipsWarningItem]
    sessions_count: int


class RipsExportSummary(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    period_year: int
    period_month: int
    status: str
    sessions_count: int
    total_value_cop: int
    file_hash: str | None
    generated_at: datetime | None


class RipsGenerationResponse(BaseModel):
    export: RipsExportSummary
    message: str

File: psicogest/frontend/src/lib/api/rips.ts
import { request, downloadBlob } from "./client";

export interface RipsExportSummary {
  id: string;
  period_year: number;
  period_month: number;
  status: string;
  sessions_count: number;
  total_value_cop: number;
  file_hash: string | null;
  generated_at: string | null;
  cuv?: string | null;
  fecha_radicacion?: string | null;
  num_factura?: string | null;
}

export interface RipsSubmitResponse {
  export_id: string;
  cuv: string | null;
  fecha_radicacion: string | null;
  result_state: boolean;
  validation_results: Array<{ [key: string]: unknown }>;
  message: string;
}

export interface RipsGenerateRequest {
  year: number;
  month: number;
}

export interface RipsValidationError {
  session_id?: string;
  field: string;
  value?: string;
  message: string;
}

export interface RipsValidationWarning {
  session_id?: string;
  field: string;
  value?: string;
  message: string;
}

export interface RipsValidateResponse {
  valid: boolean;
  errors: RipsValidationError[];
  warnings: RipsValidationWarning[];
  sessions_count: number;
}

export interface RipsGenerationResponse {
  export: RipsExportSummary;
  message: string;
}

export const ripsApi = {
  generate: (body: RipsGenerateRequest) =>
    request<RipsGenerationResponse>("POST", "/rips/generate", body),
  validate: (body: RipsGenerateRequest) =>
    request<RipsValidateResponse>("POST", "/rips/validate", body),
  list: (limit?: number) => {
    const q = new URLSearchParams();
    if (limit) q.set("limit", String(limit));
    return request<RipsExportSummary[]>("GET", `/rips?${q}`);
  },
  get: (id: string) =>
    request<RipsExportSummary>("GET", `/rips/${id}`),
  download: (id: string) =>
    downloadBlob("GET", `/rips/${id}/download`),
  submit: (id: string, body?: { num_factura?: string; xml_fev_b64?: string }) =>
    request<RipsSubmitResponse>("POST", `/rips/${id}/submit`, body ?? {}),
};
